Add phone number input to PatientSearchScreen

vinaysshenoy opened the pull request:
Pivotal Story Link: https://www.pivotaltracker.com/story/show/167343636

sanchita commented on app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsScreen.kt:
```diff
@@ -75,13 +77,13 @@ class PatientSearchResultsScreen(context: Context, attrs: AttributeSet) : Relati
+  private fun extractPatientName(criteria: PatientSearchCriteria): String {
+    return when (criteria) {
+      is Name -> criteria.patientName
+      is PhoneNumber -> TODO("not yet implemented")
```
Please implement this.

vinaysshenoy replied:
Whoops, good catch! This will be fixed when https://www.pivotaltracker.com/story/show/167343744 is picked up. For now, I will return an empty string so that the app does not crash.

sanchita commented on app/src/main/java/org/simple/clinic/search/results/PatientSearchResultsScreen.kt:
```diff
@@ -75,13 +77,13 @@ class PatientSearchResultsScreen(context: Context, attrs: AttributeSet) : Relati
-    return when (searchPatientInput) {
-      is SearchPatientInput.Name -> searchPatientInput.searchText
-      is SearchPatientInput.PhoneNumber -> TODO("not yet implemented")
+  private fun extractPatientName(criteria: PatientSearchCriteria): String {
```
Rename this appropriately.

vinaysshenoy replied:
This was a temporary "fix" because we needed to pass in a name to the `RegisterNewPatient` event. This entire function will be removed once https://www.pivotaltracker.com/story/show/167343744 is done. I'd rather not rename a function which will be removed in an upcoming PR.

I have added a `TODO` that explains the same.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt:
```diff
@@ -36,17 +36,17 @@ class PatientSearchScreenController @Inject constructor() : ObservableTransforme
+            if (text.isBlank()) {
+              errors += INPUT_EMPTY
             }
             SearchQueryValidated(errors)
```
Yeah, earlier this had name + phone number + age which were contributing their own errors. When we removed them, we didn't remove the list which was accumulating the errors. I'll remove this.

vinaysshenoy commented on app/src/main/java/org/simple/clinic/search/PatientSearchScreenController.kt:
```diff
@@ -75,10 +75,6 @@ class PatientSearchScreenController @Inject constructor() : ObservableTransforme
-        .map { it.text.trim() }
-
     val validationErrors = events
         .ofType<SearchQueryValidated>()
```
I don't think we should remove multiple errors from the `PatientSearchValidationError` at the moment. There is currently no validation done on the phone number at the moment, but we can't say there will never be one in the future.

However, I do think we can abstract the fact that the input is valid means that the validation errors are empty by converting that chain into an `Observable<Boolean>`. I'll try this out.

vinaysshenoy replied:
Tried this, looks a lot cleaner now!

vinaysshenoy commented:
@sanchita I have also removed the `TextInput` type that was introduced internally in `PatientSearchScreenController` and changed it to generate the `PatientSearchCriteria` directly.
